### PR TITLE
Allow desktops to be unverified

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Previously once a desktop was flagged as "verified", it was impossible to unset the flag. This lead to some odd behaviour such as:

```
$ bin/desktop avail
┌───────┬────────────────────────────────────────────────────────────┬──────────┐
│ Name  │ Summary                                                    │ State    │
├───────┼────────────────────────────────────────────────────────────┼──────────┤
│ xterm │ Minimal desktop environment with an xterm terminal window. │ Verified │
│       │                                                            │          │
│       │  > https://invisible-island.net/xterm/xterm.html           │          │
│       │                                                            │          │
└───────┴────────────────────────────────────────────────────────────┴──────────┘
$ yum remove xterm
$ bin//desktop verify xterm --force
Verifying desktop type xterm:

   > ✅ Flight Desktop prerequisites 
   > ✅ Package: xorg-x11-server-utils 
   > ❌ Package: xterm 

Desktop type xterm has missing prerequisites:

 * Package: xterm

Before this desktop type can be used, it must be prepared using the
'prepare' command, i.e.:
 $ bin/desktop avail
┌───────┬────────────────────────────────────────────────────────────┬──────────┐
│ Name  │ Summary                                                    │ State    │
├───────┼────────────────────────────────────────────────────────────┼──────────┤
│ xterm │ Minimal desktop environment with an xterm terminal window. │ Verified │
│       │                                                            │          │
│       │  > https://invisible-island.net/xterm/xterm.html           │          │
│       │                                                            │          │
└───────┴────────────────────────────────────────────────────────────┴──────────┘
```

Now after the `verify --force` command is ran, the desktop goes back the the `Unverified` state.

### CODE REVIEW

**Comment by benarmston** [comment]
There is still an issue for non-root users.

```sh
[flight@gateway1 (ben1) ~]$ sudo su  - root bash -c 'flight start; flight desktop avail'
┌──────────┬──────────────────────────────────────────────────────────────────────┬────────────┐
│ Name     │ Summary                                                              │ State      │
├──────────┼──────────────────────────────────────────────────────────────────────┼────────────┤
│ chrome   │ Full-screen Google Chrome browser session.                           │ Unverified │
│          │                                                                      │            │
│          │  > https://www.google.com/chrome/                                    │            │
│          │                                                                      │            │
│ gnome    │ GNOME v3, a free and open-source desktop environment for Unix-like   │ Verified   │
│          │ operating systems.                                                   │            │
│          │                                                                      │            │
│          │  > https://www.gnome.org/                                            │            │
│          │                                                                      │            │
│ kde      │ KDE Plasma Desktop (KDE 4). Plasma is KDE's desktop environment.     │ Verified   │
│          │ Simple by default, powerful when needed.                             │            │
│          │                                                                      │            │
│          │  > https://kde.org/                                                  │            │
│          │                                                                      │            │
│ terminal │ Preconfigured terminal for Flight HPC environments.                  │ Verified   │
│          │                                                                      │            │
│ xfce     │ Xfce is a lightweight desktop environment for UNIX-like operating    │ Verified   │
│          │ systems. It aims to be fast and low on system resources, while still │            │
│          │ being visually appealing and user friendly.                          │            │
│          │                                                                      │            │
│          │  > https://xfce.org/                                                 │            │
│          │                                                                      │            │
│ xterm    │ Minimal desktop environment with an xterm terminal window.           │ Verified   │
│          │                                                                      │            │
│          │  > https://invisible-island.net/xterm/xterm.html                     │            │
│          │                                                                      │            │
└──────────┴──────────────────────────────────────────────────────────────────────┴────────────┘
[flight@gateway1 (ben1) ~]$ sudo su  - root bash -c 'flight start; flight desktop verify --force xterm'
Verifying desktop type xterm:

   > ✅ Flight Desktop prerequisites  
   > ✅ Package: xorg-x11-server-utils  
   > ✅ Package: xterm  

Desktop type xterm has been verified.

[flight@gateway1 (ben1) ~]$ flight desktop verify --force xterm
Verifying desktop type xterm:

   > ✅ Flight Desktop prerequisites  
   > ✅ Package: xorg-x11-server-utils  
   > ✅ Package: xterm  

Desktop type xterm has been verified.

[flight@gateway1 (ben1) ~]$ flight desktop avail
┌──────────┬──────────────────────────────────────────────────────────────────────┬────────────┐
│ Name     │ Summary                                                              │ State      │
├──────────┼──────────────────────────────────────────────────────────────────────┼────────────┤
│ chrome   │ Full-screen Google Chrome browser session.                           │ Unverified │
│          │                                                                      │            │
│          │  > https://www.google.com/chrome/                                    │            │
│          │                                                                      │            │
│ gnome    │ GNOME v3, a free and open-source desktop environment for Unix-like   │ Verified   │
│          │ operating systems.                                                   │            │
│          │                                                                      │            │
│          │  > https://www.gnome.org/                                            │            │
│          │                                                                      │            │
│ kde      │ KDE Plasma Desktop (KDE 4). Plasma is KDE's desktop environment.     │ Verified   │
│          │ Simple by default, powerful when needed.                             │            │
│          │                                                                      │            │
│          │  > https://kde.org/                                                  │            │
│          │                                                                      │            │
│ terminal │ Preconfigured terminal for Flight HPC environments.                  │ Verified   │
│          │                                                                      │            │
│ xfce     │ Xfce is a lightweight desktop environment for UNIX-like operating    │ Verified   │
│          │ systems. It aims to be fast and low on system resources, while still │            │
│          │ being visually appealing and user friendly.                          │            │
│          │                                                                      │            │
│          │  > https://xfce.org/                                                 │            │
│          │                                                                      │            │
│ xterm    │ Minimal desktop environment with an xterm terminal window.           │ Verified   │
│          │                                                                      │            │
│          │  > https://invisible-island.net/xterm/xterm.html                     │            │
│          │                                                                      │            │
└──────────┴──────────────────────────────────────────────────────────────────────┴────────────┘
[flight@gateway1 (ben1) ~]$ sudo yum remove xterm
Loaded plugins: fastestmirror, langpacks
Resolving Dependencies
--> Running transaction check
---> Package xterm.x86_64 0:295-3.el7 will be erased
--> Finished Dependency Resolution

Dependencies Resolved

=====================================================================================================
 Package             Arch                 Version                 Repository                    Size
=====================================================================================================
Removing:
 xterm               x86_64               295-3.el7               @centos-7-base               1.2 M

Transaction Summary
=====================================================================================================
Remove  1 Package

Installed size: 1.2 M
Is this ok [y/N]: y
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Erasing    : xterm-295-3.el7.x86_64                                                            1/1 
  Verifying  : xterm-295-3.el7.x86_64                                                            1/1 

Removed:
  xterm.x86_64 0:295-3.el7                                                                           

Complete!
[flight@gateway1 (ben1) ~]$ flight desktop avail
┌──────────┬──────────────────────────────────────────────────────────────────────┬────────────┐
│ Name     │ Summary                                                              │ State      │
├──────────┼──────────────────────────────────────────────────────────────────────┼────────────┤
│ chrome   │ Full-screen Google Chrome browser session.                           │ Unverified │
│          │                                                                      │            │
│          │  > https://www.google.com/chrome/                                    │            │
│          │                                                                      │            │
│ gnome    │ GNOME v3, a free and open-source desktop environment for Unix-like   │ Verified   │
│          │ operating systems.                                                   │            │
│          │                                                                      │            │
│          │  > https://www.gnome.org/                                            │            │
│          │                                                                      │            │
│ kde      │ KDE Plasma Desktop (KDE 4). Plasma is KDE's desktop environment.     │ Verified   │
│          │ Simple by default, powerful when needed.                             │            │
│          │                                                                      │            │
│          │  > https://kde.org/                                                  │            │
│          │                                                                      │            │
│ terminal │ Preconfigured terminal for Flight HPC environments.                  │ Verified   │
│          │                                                                      │            │
│ xfce     │ Xfce is a lightweight desktop environment for UNIX-like operating    │ Verified   │
│          │ systems. It aims to be fast and low on system resources, while still │            │
│          │ being visually appealing and user friendly.                          │            │
│          │                                                                      │            │
│          │  > https://xfce.org/                                                 │            │
│          │                                                                      │            │
│ xterm    │ Minimal desktop environment with an xterm terminal window.           │ Verified   │
│          │                                                                      │            │
│          │  > https://invisible-island.net/xterm/xterm.html                     │            │
│          │                                                                      │            │
└──────────┴──────────────────────────────────────────────────────────────────────┴────────────┘
[flight@gateway1 (ben1) ~]$ flight desktop verify --force xterm
Verifying desktop type xterm:

   > ✅ Flight Desktop prerequisites  
   > ✅ Package: xorg-x11-server-utils  
   > ❌ Package: xterm  

Desktop type xterm has missing prerequisites:

 * Package: xterm

Before this desktop type can be used, it must be prepared by your
cluster administrator using the 'prepare' command, i.e.:

  flight desktop prepare xterm

[flight@gateway1 (ben1) ~]$ flight desktop avail
┌──────────┬──────────────────────────────────────────────────────────────────────┬────────────┐
│ Name     │ Summary                                                              │ State      │
├──────────┼──────────────────────────────────────────────────────────────────────┼────────────┤
│ chrome   │ Full-screen Google Chrome browser session.                           │ Unverified │
│          │                                                                      │            │
│          │  > https://www.google.com/chrome/                                    │            │
│          │                                                                      │            │
│ gnome    │ GNOME v3, a free and open-source desktop environment for Unix-like   │ Verified   │
│          │ operating systems.                                                   │            │
│          │                                                                      │            │
│          │  > https://www.gnome.org/                                            │            │
│          │                                                                      │            │
│ kde      │ KDE Plasma Desktop (KDE 4). Plasma is KDE's desktop environment.     │ Verified   │
│          │ Simple by default, powerful when needed.                             │            │
│          │                                                                      │            │
│          │  > https://kde.org/                                                  │            │
│          │                                                                      │            │
│ terminal │ Preconfigured terminal for Flight HPC environments.                  │ Verified   │
│          │                                                                      │            │
│ xfce     │ Xfce is a lightweight desktop environment for UNIX-like operating    │ Verified   │
│          │ systems. It aims to be fast and low on system resources, while still │            │
│          │ being visually appealing and user friendly.                          │            │
│          │                                                                      │            │
│          │  > https://xfce.org/                                                 │            │
│          │                                                                      │            │
│ xterm    │ Minimal desktop environment with an xterm terminal window.           │ Verified   │
│          │                                                                      │            │
│          │  > https://invisible-island.net/xterm/xterm.html                     │            │
│          │                                                                      │            │
└──────────┴──────────────────────────────────────────────────────────────────────┴────────────┘
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ 
[flight@gateway1 (ben1) ~]$ sudo su  - root bash -c 'flight start; flight desktop verify --force xterm'
Verifying desktop type xterm:

   > ✅ Flight Desktop prerequisites  
   > ✅ Package: xorg-x11-server-utils  
   > ❌ Package: xterm  

Desktop type xterm has missing prerequisites:

 * Package: xterm

Before this desktop type can be used, it must be prepared using the
'prepare' command, i.e.:

  flight desktop prepare xterm

[flight@gateway1 (ben1) ~]$ flight desktop avail
┌──────────┬──────────────────────────────────────────────────────────────────────┬────────────┐
│ Name     │ Summary                                                              │ State      │
├──────────┼──────────────────────────────────────────────────────────────────────┼────────────┤
│ chrome   │ Full-screen Google Chrome browser session.                           │ Unverified │
│          │                                                                      │            │
│          │  > https://www.google.com/chrome/                                    │            │
│          │                                                                      │            │
│ gnome    │ GNOME v3, a free and open-source desktop environment for Unix-like   │ Verified   │
│          │ operating systems.                                                   │            │
│          │                                                                      │            │
│          │  > https://www.gnome.org/                                            │            │
│          │                                                                      │            │
│ kde      │ KDE Plasma Desktop (KDE 4). Plasma is KDE's desktop environment.     │ Verified   │
│          │ Simple by default, powerful when needed.                             │            │
│          │                                                                      │            │
│          │  > https://kde.org/                                                  │            │
│          │                                                                      │            │
│ terminal │ Preconfigured terminal for Flight HPC environments.                  │ Verified   │
│          │                                                                      │            │
│ xfce     │ Xfce is a lightweight desktop environment for UNIX-like operating    │ Verified   │
│          │ systems. It aims to be fast and low on system resources, while still │            │
│          │ being visually appealing and user friendly.                          │            │
│          │                                                                      │            │
│          │  > https://xfce.org/                                                 │            │
│          │                                                                      │            │
│ xterm    │ Minimal desktop environment with an xterm terminal window.           │ Verified   │
│          │                                                                      │            │
│          │  > https://invisible-island.net/xterm/xterm.html                     │            │
│          │                                                                      │            │
└──────────┴──────────────────────────────────────────────────────────────────────┴────────────┘
[flight@gateway1 (ben1) ~]$ sudo su  - root bash -c 'flight start; flight desktop avail'
┌──────────┬──────────────────────────────────────────────────────────────────────┬────────────┐
│ Name     │ Summary                                                              │ State      │
├──────────┼──────────────────────────────────────────────────────────────────────┼────────────┤
│ chrome   │ Full-screen Google Chrome browser session.                           │ Unverified │
│          │                                                                      │            │
│          │  > https://www.google.com/chrome/                                    │            │
│          │                                                                      │            │
│ gnome    │ GNOME v3, a free and open-source desktop environment for Unix-like   │ Verified   │
│          │ operating systems.                                                   │            │
│          │                                                                      │            │
│          │  > https://www.gnome.org/                                            │            │
│          │                                                                      │            │
│ kde      │ KDE Plasma Desktop (KDE 4). Plasma is KDE's desktop environment.     │ Verified   │
│          │ Simple by default, powerful when needed.                             │            │
│          │                                                                      │            │
│          │  > https://kde.org/                                                  │            │
│          │                                                                      │            │
│ terminal │ Preconfigured terminal for Flight HPC environments.                  │ Verified   │
│          │                                                                      │            │
│ xfce     │ Xfce is a lightweight desktop environment for UNIX-like operating    │ Verified   │
│          │ systems. It aims to be fast and low on system resources, while still │            │
│          │ being visually appealing and user friendly.                          │            │
│          │                                                                      │            │
│          │  > https://xfce.org/                                                 │            │
│          │                                                                      │            │
│ xterm    │ Minimal desktop environment with an xterm terminal window.           │ Verified   │
│          │                                                                      │            │
│          │  > https://invisible-island.net/xterm/xterm.html                     │            │
│          │                                                                      │            │
└──────────┴──────────────────────────────────────────────────────────────────────┴────────────┘

```

**Comment by WilliamMcCumstie** [comment]
On review there has been some confusion around the versioning. This output doesn't look quite right and I can't reproduce it exactly. Due to the uncertainty around the version, I won't be directly addressing your previous comment.

However there is a usability issue surrounding system vs user state files, which I will push shortly

File: lib/desktop/type.rb [hunk 1]
```diff
@@ -127,18 +127,32 @@ def supports_host_arch?
...
+      return false if paths.empty?
+
+      # Find the last updated state file
+      path = if paths.length == 1
+               paths.first
+             elsif File.ctime(paths.first) < File.ctime(paths.last)
+               paths.last
+             else
+               paths.first
+             end
```
Comment (benarmston) [inline lib/desktop/type.rb:144]: How about

```
path = paths.sort_by { |p| File.ctime(p) }.last
```

**Review by benarmston — CHANGES_REQUESTED** [review]
I'll try to reproduce the issue with `flight-desktop-1.5.0~c1-1`.  The latest commit looks like it should do the right thing.  Determining the correct path to use could be done more nicely, see review comment.

**Comment by benarmston** [comment]
`flight-desktop-1.5.0~c1-1` fixes the issue.

**Comment by benarmston** [comment]
I've made the change that I suggested to determining the state file to use.  This LGTM.

**Pull request merged**